### DISCUSSION TITLE
Detecting license plate only when the motorcycle is facing backward

### DISCUSSION THREAD

**Discussion by neliel123** [ultralytics/ultralytics#13387] [Q&A]
I have just started with YoloV8 and I am trying to create a license plate detector on
motorcycles.

Usually in my country, the motorcycle plate is situated at the back of each motorcycle
so what I did was to create a custom dataset of motorcycles with their license plate.

Example from pexels 
![image](https://github.com/ultralytics/ultralytics/assets/69466026/5a1b1730-b202-4387-90ef-2b92af8df698)

Below is the code that I am experimenting and it works.

What I am doing is detecting the motorcycle first then loading my custom license plate detector
and use easyocr to extract the text.

It works for some reason.

```
import cv2
import torch
from ultralytics import YOLO
import easyocr
from google.colab.patches import cv2_imshow

motorcycle_model = YOLO('yolov8n.pt')  

# Load the custom YOLO model for license plate detection
license_plate_model = YOLO('/content/gdrive/MyDrive/runs_motorcycle_plate/detect/train/weights/last.pt')  # Adjust the model path as necessary

reader = easyocr.Reader(['en'])

# Function to detect motorcycles and then license plates
def detect_motorcycles_and_plates(image_path):
    # Load the image
    image = cv2.imread(image_path)
    
    # Perform inference to detect motorcycles
    motorcycle_results = motorcycle_model(image)
    
    print(motorcycle_model.names)
    # Loop over the detections and extract motorcycles
    for result in motorcycle_results:
        boxes = result.boxes
        for box in boxes:
            cls = int(box.cls)
            if motorcycle_model.names[cls] == 'motorcycle':  # Filter for 'motorcycle' class
                x1, y1, x2, y2 = map(int, box.xyxy[0])
                motorcycle_crop = image[y1:y2, x1:x2]
                
                # Perform inference to detect license plates on the cropped motorcycle image
                plate_results = license_plate_model(motorcycle_crop)
                print(f"Length Plate Detections: {len(plate_results)}")
                
                # Loop over the detections and extract license plates
                for plate_result in plate_results:
                    plate_boxes = plate_result.boxes
                    print(f"license_plate_model.names :: {license_plate_model.names}")
                    for plate_box in plate_boxes:
                        plate_cls = int(plate_box.cls)
                        if license_plate_model.names[plate_cls] == 'plate_number':  # Filter for 'plate_number' class
                            px1, py1, px2, py2 = map(int, plate_box.xyxy[0])
                            license_plate_crop = motorcycle_crop[py1:py2, px1:px2]
                            
                            # Decode the license plate using EasyOCR
                            ocr_result = reader.readtext(license_plate_crop)
                            for (bbox, text, prob) in ocr_result:
                                print(f'Detected license plate text: {text} with confidence: {prob}')
                                
                                # Optionally draw the OCR result on the image
                                cv2.rectangle(motorcycle_crop, (px1, py1), (px2, py2), (0, 255, 0), 2)
                                cv2.putText(motorcycle_crop, text, (px1, py1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
                
                # Display the motorcycle crop with detected license plate
                # cv2.imshow('Motorcycle with License Plate', motorcycle_crop)
                cv2_imshow(motorcycle_crop)
                cv2.waitKey(0)
                
    # Release the OpenCV window
    cv2.destroyAllWindows()

detect_motorcycles_and_plates('/content/gdrive/MyDrive/test_data/445369253_823112873035331_1818535960009901237_n.jpg')
```

However, the problem occurs when motorcycles are detected in sidewards with the camera
and if a person holds a bag with some text in it then it is detected by the system
as a license plate.

I am thinking if I should detect if the motorcycle is facing the camera or its back
is facing. If it is facing backward then this is the only time that I could
run my license plate model.

Question, is this a viable approach or is this possible? And can I detect if a motorcycle is facing me or if its
back is being shown? Or is there a much better alternative?

I am using Google Colab.  Thanks for the help.

**Comment by pderrenger** [comment]
@neliel123 hello! It sounds like you've made great progress with your license plate detection project using YOLOv8. Your approach to detect the motorcycle orientation before running the license plate model is indeed a viable strategy. This can help reduce false positives from unrelated text on objects like bags.

To determine the orientation of the motorcycle, you might consider training your model to recognize different angles of motorcycles, such as 'front', 'back', 'side', etc. This way, you can condition the license plate detection to activate only when the 'back' class is detected.

Another alternative could be to enhance your dataset with more varied examples of motorcycles from different angles and with various occlusions (like bags) to improve the model's robustness in distinguishing between actual license plates and similar-looking objects.

Using Google Colab is a good choice for leveraging GPUs for training and inference, ensuring you can iterate quickly on your model training.

Keep up the good work, and don't hesitate to reach out if you have more questions! 🚀
